[IMP] attachment_preview: improve performance for large files with filenames

tremlin opened the pull request:
This PR rewrites `get_binary_extension` so that it uses a first pass to determine the extension from the filename only whenever possible. This avoids to load big files into memory if only the filename is needed.

Commit:
[IMP] rewrite get_binary_extension so that it uses a first pass to determine the extension from the filename only

pedrobaeza commented:
Are you sure magic reads whole file? AFAIK, it only reads first bytes.

tremlin commented:
Magic usually reads only a few bytes, however not always at the beginning (http://linux.die.net/man/5/magic). The core problem is not really with magic but with Odoo, which provides AFAIK no fast random file access.

pedrobaeza commented:
OK, I let @hbrunn to give his opinion on this.

hbrunn commented on attachment_preview/model/ir_attachment.py:
```diff
@@ -32,35 +32,47 @@ def get_binary_extension(
+                    result[this.id] = False
+        # Seconds pass for all attachments which have to be loaded fully
+        #  to get the extension from the content
+        ids_to_browse = [_id for _id in ids_to_browse if _id not in result]
```
how can any id be left here in the happy flow? That is, don't you effectively disable mime type tests?

hbrunn commented:
as for random access: Since version 8, odoo stores attachments on the filesystem by default. So I think your patch would be a lot better if you check for this the way core does it in https://github.com/OCA/OCB/blob/8.0/openerp/addons/base/ir/ir_attachment.py#L161 and just read from the file system if you can via https://github.com/OCA/OCB/blob/8.0/openerp/addons/base/ir/ir_attachment.py#L100. This should solve your problem while still using the magic and mimetype logic

Commit:
[IMP] apply file directly to local file if storage is 'file'

tremlin commented:
Thanks for the review. The logic change was not intended this way and I've corrected it. That is, if no extension could be obtained from the filename, magic will be tried on the file.

Regarding passing the storage filename to magic directly: It works only for attachments in Odoo 8 and not for binary fields in general (only Odoo 9 has the attachment attribute of `binary`). I implemented the change in a second commit, but I am not sure how helpful it is in general.
The application that I had in mind is exactly the opposite: arbitrary binary fields which are stored in the database :)

hbrunn commented on attachment_preview/model/ir_attachment.py:
```diff
@@ -32,35 +32,50 @@ def get_binary_extension(
+        if filename_field:
+            context_bin_size = dict(context or {})
+            context_bin_size['bin_size'] = True
+            ids_to_browse = ids if isinstance(ids, collections.Iterable) \
```
move this out of the conditional to avoid possible crashes in line 55 below

tremlin replied:
Thanks. Don't change code after testing...

hbrunn commented on attachment_preview/model/ir_attachment.py:
```diff
@@ -32,35 +32,50 @@ def get_binary_extension(
+                mimetype.split(';')[0], strict=False)
+            result[this.id] = extension
+        for _id in result:
+            result[_id] = (extension or '').lstrip('.').lower()
```
was this here before? This won't crash most of the time, but assign the last extension we got to all requested records.

Maybe add a test for this function so that we're sure it behaves correctly?

Commit:
[FIX] fix bugs and add logging

tremlin commented:
Sorry for the hot action yesterday.
I tested the now corrected code with both cases (extension from filename and file magic).

pedrobaeza reviewed (approved):
Following Holger's expert approval.
